packages/grafana-ui/src/components/Chart/TooltipContainer.tsx

```diff
@@ -20,18 +20,37 @@ export const TooltipContainer: React.FC<TooltipContainerProps> = ({
 }) => {
   const theme = useTheme();
   const tooltipRef = useRef<HTMLDivElement>(null);
+  const tooltipMeasurementRef = useRef<Dimensions2D>({ width: 0, height: 0 });
   const { width, height } = useWindowSize();
   const [placement, setPlacement] = useState({
     x: positionX + offsetX,
     y: positionY + offsetY,
   });
 
+  const resizeObserver = new ResizeObserver((entries) => {
+    for (let entry of entries) {
+      const tW = Math.floor(entry.contentRect.width + 2 * 8); //  adding padding until Safari supports borderBoxSize
+      const tH = Math.floor(entry.contentRect.height + 2 * 8);
+
+      if (tooltipMeasurementRef.current.width !== tW || tooltipMeasurementRef.current.height !== tH) {
+        tooltipMeasurementRef.current = {
+          width: tW,
+          height: tH,
+        };
+      }
+    }
+  });
+
+  if (tooltipRef.current) {
+    resizeObserver.observe(tooltipRef.current);
+  }
+
```

Comment:
i noticed the garbage collector suddenly kicked into overdrive, and then realized all of this is getting called and recreated on every tooltip move. let's make sure this whole block is memoized and only called once.

Reply:
good catch!